api: Improve session lifecycle handling

victorges opened the pull request:
**What does this pull request do? Explain your changes. (required)**

This is a bit massive change, but it completely rewrites the way that we handle livestream
sessions and their corresponding recordings on the API. 

The explanation here started to get too big so I wrote a doc instead: 
https://lvpr.link/live-rec-fact

**Specific updates (required)**
 - Made the "session merging" fields private and deprecated
 - Removed all "write" usage of them, keeping only the "read" usage for backward compatibility on existing streams
 - Added a logic to re-use previous transcoding sessions if they are recent and haven't been used (crash-loop protection)
 - Started writing the `isActive` field on child streams ("transcoding sessions") from `stream-info-service`
 - Used that `isActive` field to trigger recording webhooks and generation of VOD assets from them
 - Created new `none` state on the `recordingStatus` to differentiate in progress from failed recordings
 - Make sure to send `recording.ready` and create VOD assets when the `isActive` hadn't been updated by mapic

**How did you test each of these updates (required)**
 - Record a livestream
 - Make sure the session shows up on the stream page
 - Make sure it has the correct length
 - Make sure it has recording link and that it works
 - Make sure a VOD asset is generated from the recording
 - Kill `MistProcLivepeer` mid-stream
 - Make sure a new session shows up and all recordings are available in the end

**Does this pull request close any open issues?**
Implements API-7

**Screenshots (optional)**
![image](https://user-images.githubusercontent.com/1613383/223244629-f19311d2-aab3-4fe9-94e2-386ca91e8752.png)
This `n/a` is so much better than all of them being `In progress` as before :feelsgood:
We can also later change the frontend to only display the sessions that actually have something in them (but lets wait for catalyst recordings)

**Checklist**

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the **CONTRIBUTING** document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

victorges committed:
api/schema: Remove partial session fields from streams

victorges committed:
api/schema: Bring back the fields as deprecated

Next step is removing all "write" use of them

victorges committed:
api/stream: Remove logic for merging consecutive sessions

Still need to figure out the right logic for sending
recording.ready hooks though.

victorges committed:
api/stream: Fix other previous sessions logic

victorges committed:
api: Remove yet other uses of deprecated fields

victorges committed:
api/stream: Improve child streams ("sessions") lifecycle

 - Avoid creating a bunch of repeated empty sessions by reusing them if
   they are recent+unused
 - Use isActive field to generate all of the recordings properly

victorges committed:
api/cannon: Create helper for recording.ready handling

victorges committed:
api/cannon: Only drop messages that lead to runtime errors

We should retry on regular errors like networking, timeouts, etc.

victorges committed:
api/stream: Fix handling of recording OS ID and status

victorges committed:
api: Make recordings show up faster

victorges committed:
api: Send webhooks on active cleanup of inconsistent streams

victorges committed:
api/cannon: Double check that the stream stopped

victorges committed:
api/stream: Only set active preemptively for parent streams

victorges committed:
api/stream: Fix sendSetActiveHooks to use session whwere appropriate

victorges committed:
api/stream: Simplify the whole isActive logic

iameli commented on packages/api/src/webhooks/cannon.ts:
```diff
@@ -502,18 +489,44 @@ export default class WebhookCannon {
+    }
+    await this.db.stream.update(sessionId, { isActive: false });
+
+    await this.recordingToVodAsset(session, mp4Url);
```
Is there an argument to be made for doing these in the opposite order in case `recordingToVodAsset` fails for whatever reason?

victorges replied:
I'm having to do the DB update first in order to avoid generating duplicate assets for the same recording. An alternative would be creating a unique index on the `source.sessionId` field but that would seem like overkill TBH. We need to sort out our transaction game here.

iameli reviewed (approved):
LGTM to the extent I still remember how this recording workflow works 🤔

gioelecerati commented on packages/api/src/store/stream-table.ts:
```diff
@@ -41,13 +41,36 @@ export interface StreamStats {
 }
 
-export interface DBStreamFields extends StreamStats {
-  previousStats?: StreamStats;
```
Something got broken into the tests because of this field, although I cannot spot the reason, it just got moved.

victorges replied:
Fixed!

gioelecerati commented on packages/api/src/controllers/session.ts:
```diff
@@ -134,97 +134,6 @@ app.get("/", authorizer({}), async (req, res, next) => {
   res.json(sessions);
 });
 
-app.get("/migrate2", authorizer({ anyAdmin: true }), async (req, res, next) => {
```
Why these two endpoints got removed? What they were used for and by what?

victorges replied:
They were for migrating super old database schemas into old database schemas. They were probably called to exhaustion the first time they were created and were forgotten to be deleted. We can always look back for this for inspiration in case we ever want to create more migration flows like this (e.g. to migrate storage schema in assets, like we did recently last year).

gioelecerati commented on packages/api/src/controllers/stream.ts:
```diff
@@ -956,60 +921,99 @@ app.put(
+  }
+
+  for (const session of sessions) {
+    try {
```
nit: `session.id` could be put in a const here to be used in both the try block and the error message
```suggestion
    const sessionId = session.id;
    try {
```

victorges replied:
Not sure I like it that much as in the end it will just mean 1 less character to write for the same thing 😆, and I like the clarity of `session.id` which explicitly tells where it comes from

gioelecerati commented on packages/api/src/controllers/stream.ts:
```diff
@@ -956,60 +921,99 @@ app.put(
+    }
+  }
+
+  for (const session of sessions) {
```
nit: could be breakdown in two separate smaller functions for clarity
```suggestion
  for (const session of sessions) {
    try {
      await processSession(session, queue, ingest);
    } catch (err) {
      logger.error(
        `Error processing session session_id=${session.id} err=`,
        err
      );
    }
  }
}

async function processSession(
  session: DBStream,
  queue: Queue,
  ingest: string
) {
```

victorges replied:
As we are only talking about 7 lines of logic, I think it's kinda overkill to extract this helper, ChatGPT does some of those every now and then... I'm gonna do it anyway in case you also think it's more readable, not only copied from GPT 😄

victorges replied:
Done, but gave it a better name than a generic "processSession" that would only obfuscate what is done in that piece of the code. Perhaps the whole logic from `triggerSessionRecordingHooks` is more easily understandable now only by skimming through the code.

gioelecerati reviewed (approved):
LGTM with comments

victorges committed:
api: Fix test expectations not to join sessions

victorges committed:
api: Create helper for single recording.ready hook

victorges committed:
api: Make sure to update DB consistently

victorges committed:
api: A couple fixes to the logic

victorges committed:
api: Avoid updating parent streams isActive field

Bugs the dashboard as we might re-set the stream
as active after the mapic setactive hook

victorges committed:
api: Simplify throwIfEmpty spaghetti

victorges committed:
api: Re-throw unknown errors from handleRecordingChecks

victorges committed:
stream-info-service: Avoid updating lastSeen after no segments

victorges committed:
stream-info-app: Avoid updating isActive field on no segments

victorges committed:
api: Stop setting isActive from stream-info-app

It is leading to duplicate assets generated as oftentimes
stream-info-app sets isActive to true again after the
recording asset had already been generated. Instead of that,
just create every child session as active and treat the
isActive as a flag for whether the recording has been generated
or not (without a chance it is made true again).

victorges committed:
api: Add log when we ignore a session recording

Also fix tests

victorges committed:
stream-info-app: Increase delete timeout

Was seeing some sessions getting duplicated stats
from them being deleted and re-created

victorges committed:
stream-info-app: Make sure to stop saving metrics on the old sessions

victorges committed:
Revert "stream-info-app: Increase delete timeout"

This reverts commit c0667ef14be84e9b0ec5086ee55a3a67c98e7034.
